librz/core/cconfig.c

```diff
@@ -1737,6 +1737,121 @@ RZ_API bool rz_core_esil_cmd(RzAnalysisEsil *esil, const char *cmd, ut64 a1, ut6
 	return false;
 }
 
+static void config_print_node(RzConfig *cfg, RzConfigNode *node, RzCmdStateOutput *state) {
+	rz_return_if_fail(cfg && node && state);
+	char *option;
+	bool isFirst;
+	RzOutputMode mode = state->mode;
+	PJ *pj = state->d.pj;
+	RzListIter *iter;
+	char *es = NULL;
+
+	switch (mode) {
+	case RZ_OUTPUT_MODE_JSON:
+		if (rz_str_isnumber(node->value)) {
+			pj_kn(pj, node->name, rz_num_math(NULL, node->value));
+			return;
+		} else if (rz_str_is_bool(node->value)) {
+			pj_kb(pj, node->name, (node->value));
+			return;
+		} else {
+			pj_ks(pj, node->name, node->value);
+		}
+		break;
+	case RZ_OUTPUT_MODE_LONG_JSON:
+		pj_o(pj);
+		pj_ks(pj, "name", node->name);
+		if (rz_str_isnumber(node->value)) {
+			pj_kn(pj, "value", rz_num_math(NULL, node->value));
+		} else if (rz_str_is_bool(node->value)) {
+			pj_kb(pj, "value", (node->value));
+		} else {
+			pj_ks(pj, "value", node->value);
+		}
+		pj_ks(pj, "type", rz_config_node_type(node));
+		es = rz_str_escape(node->desc);
+		if (es) {
+			pj_ks(pj, "desc", es);
+			free(es);
+		}
+		pj_kb(pj, "ro", rz_config_node_is_ro(node));
+		if (!rz_list_empty(node->options)) {
+			pj_ka(pj, "options");
+			rz_list_foreach (node->options, iter, option) {
+				pj_s(pj, option);
+			}
+			pj_end(pj);
+		}
+		pj_end(pj);
+		break;
+	case RZ_OUTPUT_MODE_LONG:
+		rz_cons_printf("%s = %s %s; %s",
+			node->name, node->value,
+			rz_config_node_is_ro(node) ? "(ro)" : "",
+			node->desc);
+		if (!rz_list_empty(node->options)) {
+			isFirst = true;
+			rz_cons_printf(" [");
+			rz_list_foreach (node->options, iter, option) {
+				if (isFirst) {
+					isFirst = false;
+				} else {
+					rz_cons_printf(", ");
+				}
+				rz_cons_printf("%s", option);
+			}
+			rz_cons_printf("]");
+		}
+		rz_cons_println("");
+		break;
+	case RZ_OUTPUT_MODE_QUIET:
+		rz_cons_printf("%s=%s\n", node->name, node->value);
+		break;
+	case RZ_OUTPUT_MODE_RIZIN:
+		es = rz_cmd_escape_arg(node->value, RZ_CMD_ESCAPE_ONE_ARG);
+		rz_cons_printf("e %s=%s\n", node->name, es);
+		free(es);
+		break;
+	case RZ_OUTPUT_MODE_STANDARD:
+		rz_cons_printf("%20s: %s\n", node->name,
+			node->desc ? node->desc : "");
+		break;
+	default:
+		rz_warn_if_reached();
+		break;
+	}
+}
+
+RZ_API void rz_core_config_print_all(RzConfig *cfg, const char *str, RzCmdStateOutput *state) {
+	rz_return_if_fail(cfg);
+	RzConfigNode *node;
+	RzListIter *iter;
+	int len = 0;
+	PJ *pj = state->d.pj;
+	RzOutputMode mode = state->mode;
+
+	if (mode == RZ_OUTPUT_MODE_LONG_JSON) {
```

Comment:
@ret2libc is this still required? I believe you added it as a `rz_cmd_state_output_array_start()` (and the footer `rz_cmd_state_output_array_end()`). Is the difference of using `pj_o()` for "short JSON" is really required?

Reply:
yes in this case it is needed because there's a difference between `elj` and `elJ`. `elj` just returns a json object like:
```json
{
  "cfg.oldshell": true,
  "asm.var.sub": false,
}
```
while `elJ` returns:
```json
[
  {
    "name": "cfg.oldshell",
   "value": true,
   "type": "boolean",
  "option": ....
   "description: "...."
  },
  {
    "name": "asm.var.sub",.....
  }
]
```

Reply:
So in one case you have an array, in the other you have an object.